### PR TITLE
feat: add json ld structure data into blog pages

### PR DESCRIPTION
**Pull request by iamnamananand996**
will close this issue https://github.com/instill-ai/instill.tech/issues/495

Because

- Google provide a better way to inform the robot about your site which is JSONLD structure data. We wanted to implement this feature into our `instill.tech`.

This commit

- Have basic `JSONLD` for blog pages currently.

**exmaple** 
https://github.com/instill-ai/instill.tech/issues/495#issuecomment-1466608891

### CODE REVIEW

File: src/pages/blog/[...path].tsx [hunk 1]
```diff
@@ -143,6 +142,48 @@ const BlogPage: FC<BlogPageProps> & {
...
+        headline: blogMeta.title,
+        image: process.env.NEXT_PUBLIC_BASE_URL + blogMeta.themeImgSrc,
+        url: process.env.NEXT_PUBLIC_BASE_URL + "/blog/" + blogMeta.slug,
+        dateCreated: blogMeta.publishedOn,
+        dateModified: blogMeta.commit.lastEditedTime,
+        datePublished: blogMeta.publishedOn,
+        inLanguage: blogMeta.lang,
+        isFamilyFriendly: true,
+        keywords: [blogMeta.slug],
+        articleBody: blogMeta.description,
```
Comment (EiffelFly) [inline src/pages/blog/[...path].tsx:161]: @iamnamananand996 From the definition https://schema.org/articleBody articleBody means the actual body of the article, in this case we can put the whole source string here. You could export another props `source` from the getStaticProps which is this [line](https://github.com/instill-ai/instill.tech/blob/366a1507d38ea5d65867ad7dc7d176dc3054415b/src/pages/blog/%5B...path%5D.tsx#L77) And directly put the source into articleBody.

File: src/pages/blog/[...path].tsx [hunk 2]
```diff
@@ -143,6 +142,48 @@ const BlogPage: FC<BlogPageProps> & {
...
+        "@type": "BlogPosting",
+        headline: blogMeta.title,
+        image: process.env.NEXT_PUBLIC_BASE_URL + blogMeta.themeImgSrc,
+        url: process.env.NEXT_PUBLIC_BASE_URL + "/blog/" + blogMeta.slug,
+        dateCreated: blogMeta.publishedOn,
+        dateModified: blogMeta.commit.lastEditedTime,
+        datePublished: blogMeta.publishedOn,
+        inLanguage: blogMeta.lang,
+        isFamilyFriendly: true,
+        keywords: [blogMeta.slug],
```
Comment (EiffelFly) [inline src/pages/blog/[...path].tsx:160]: We will have keywords frontmatter in our blogPost in the future. Let's leave this one as blank array this PR

```suggestion
        keywords: [],
```

File: src/pages/blog/[...path].tsx [hunk 3]
```diff
@@ -143,6 +142,48 @@ const BlogPage: FC<BlogPageProps> & {
...
+        author: {
+          "@type": "Person",
+          name: blogMeta.author,
+          url: blogMeta.authorGitHubUrl,
+        },
+        publisher: {
+          "@type": "Person",
+          name: blogMeta.author,
+          url: blogMeta.commit.authorAvatarUrl,
+        },
```
Comment (EiffelFly) [inline src/pages/blog/[...path].tsx:171]: @iamnamananand996 The published should be our organization

```suggestion
        publisher: {
          "@type": "Organization",
          name: "Instill AI",
          url: process.env.NEXT_PUBLIC_BASE_URL,
        },
```

File: src/pages/blog/[...path].tsx [hunk 4]
```diff
@@ -143,6 +142,48 @@ const BlogPage: FC<BlogPageProps> & {
     );
   }, [articles, blogMeta]);
 
+  const getJsonLd = (
+    blogMeta: Nullable<BlogArticleMeta>
+  ): Nullable<BlogArticleJsonLD | object> => {
```
Comment (EiffelFly) [inline src/pages/blog/[...path].tsx:147]: The type is a bit weird, object is not a concrete type. I would suggest that we make it cleaner

```suggestion
  const getJsonLd = (
    blogMeta: Nullable<BlogArticleMeta>
  ): Nullable<BlogArticleJsonLD> => {
```

@iamnamananand996 Please refactor function and other related functions to adapt this change

**Review by EiffelFly — COMMENTED** [review]
@iamnamananand996 The naming of BlogArticleJsonLD is good. You are following up our approach!! Good job. I had left some other suggestion, please check them

**Comment by iamnamananand996** [comment]
updated the suggested request changes.
cc - @EiffelFly

File: src/components/ui/PageHead.tsx [hunk 5]
```diff
@@ -11,6 +16,7 @@ export type PageHeadProps = {
   additionMeta: Nullable<ReactElement>;
   currentArticleMeta: Nullable<TutorialMeta | BlogArticleMeta>;
   commitMeta: Nullable<CommitMeta>;
+  jsonLd?: Nullable<BlogArticleJsonLD>;
```
Comment (EiffelFly) [inline src/components/ui/PageHead.tsx:19]: @iamnamananand996 In our codebase we want to enforce the explicit props policy. So we try to force every prop not take undefind as their type. 

```suggestion
  jsonLd: Nullable<BlogArticleJsonLD>;
```

File: src/components/ui/PageHead.tsx [hunk 6]
```diff
@@ -210,6 +217,10 @@ export const PageHead = ({
             <link rel="shortcut icon" href="/instill-ai-favicon.ico" />
           </>
         )}
+
+        {pageType === "blog" && (
+          <script type="application/ld+json">{JSON.stringify(jsonLd)}</script>
+        )}
```
Comment (EiffelFly) [inline src/components/ui/PageHead.tsx:223]: @iamnamananand996 We need to further check whether JSONLD is present

```suggestion

        {pageType === "blog" && jsonLd && (
          <script type="application/ld+json">{JSON.stringify(jsonLd)}</script>
        )}
```

**Review by EiffelFly — COMMENTED** [review]
@iamnamananand996 I had further added some review, please help me resolve them

**Comment by iamnamananand996** [comment]
@EiffelFly , updated suggested changes now

**Comment by iamnamananand996** [comment]
@EiffelFly , fix the above build failure.

**Comment by EiffelFly** [comment]
@iamnamananand996 We are currently not allow the undefined value in this component, please alter all the page that has PageHead to set jsonLd prop as null to fix this issue.

**Comment by iamnamananand996** [comment]
@EiffelFly, updated all the page that has `<PageHead />` to set `jsonLd` prop as `null`

**Pull request merged**